### PR TITLE
refactor(storage): deserialize output columns only

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Deserialize only the required column in StorageTable's get and iter.
- Fix a bug in `ColumnAwareSerde::new` when calculating `column_with_default`.

Find more details in
#https://github.com/risingwavelabs/risingwave/issues/16763#issuecomment-2123882211

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/row_serde/value_serde.rs [hunk 1]
```diff
@@ -140,6 +141,21 @@ impl ValueRowSerde for ColumnAwareSerde {
...
+        Self(inner)
+    }
+}
+
+impl ValueRowDeserializer for PartialRowDeserializer {
+    fn deserialize(&self, encoded_bytes: &[u8]) -> Result<Vec<Datum>> {
+        self.0.deserialize(encoded_bytes)
+    }
+}
+
```
Comment (BugenZhao) [inline src/storage/src/row_serde/value_serde.rs:158]: This looks to behave the exactly same as `Deserializer`. 🤔 How could it be partial (or the original one not be partial)?

Comment (zwang28) [reply src/storage/src/row_serde/value_serde.rs:158]: yes I should get rid of this dummy type.

File: src/storage/src/table/batch_table/storage_table.rs [hunk 2]
```diff
@@ -264,6 +273,32 @@ impl<S: StateStore> StorageTableInner<S, EitherSerde> {
...
+            let output_row_in_value_output_indices_dedup = value_output_indices
+                .iter()
+                .map(|&di| {
+                    value_output_indices_dedup
+                        .iter()
+                        .position(|&pi| di == pi)
+                        .unwrap()
+                })
+                .collect_vec();
+            let partial_row_serde = ColumnAwareSerde::new(
```
Comment (hzxa21) [inline src/storage/src/table/batch_table/storage_table.rs:292]: For `row_serde`, we use different Serde implementation depending on whether `versioned` is true. Do we need to do the same trick here?

File: src/storage/src/table/batch_table/storage_table.rs [hunk 3]
```diff
@@ -83,6 +85,13 @@ pub struct StorageTableInner<S: StateStore, SD: ValueRowSerde> {
 
     /// Row deserializer to deserialize the whole value in storage to a row.
     row_serde: Arc<SD>,
+    /// The `Deserializer` deserializes the columns specified in `value_output_indices`.
+    /// The `Deserializer` ensures there are no duplicate columns because `ColumnAwareSerde` disallows them.
+    /// The `ColumnMapping` restores all columns specified in `value_output_indices`.
+    partial_row_deserializer: Option<(
+        Arc<column_aware_row_encoding::Deserializer>,
+        Arc<ColumnMapping>,
```
Comment (hzxa21) [inline src/storage/src/table/batch_table/storage_table.rs:93]: Why do we need both `row_serde` and `partial_row_deserializer`? Is there a case when we need to access the full row even though the `value_output_indices` is set to only ouputing a subset of columns in value?

File: src/storage/src/table/batch_table/storage_table.rs (hunk 2 above)
Comment (zwang28) [reply src/storage/src/table/batch_table/storage_table.rs:292]: It is already. For !versioned, partial_row_deserializer is not available.
`let partial_row_deserializer = if versioned`

Comment (hzxa21) [reply src/storage/src/table/batch_table/storage_table.rs:292]: Oh my bad. I missed that if check at the beginning. Thanks for the clarification.

File: src/storage/src/table/batch_table/storage_table.rs (hunk 3 above)
Comment (zwang28) [reply src/storage/src/table/batch_table/storage_table.rs:93]: I kept both for potential usage in the future. 
I'll unify them.

File: src/storage/src/table/batch_table/storage_table.rs [hunk 4]
```diff
@@ -246,29 +247,46 @@ impl<S: StateStore> StorageTableInner<S, EitherSerde> {
...
-
-        let row_serde = {
+        let (row_serde, mapping) = {
             if versioned {
-                ColumnAwareSerde::new(value_indices.into(), table_columns.into()).into()
+                let value_output_indices_dedup = value_output_indices
+                    .iter()
+                    .sorted()
+                    .dedup()
+                    .copied()
```
Comment (BugenZhao) [inline src/storage/src/table/batch_table/storage_table.rs:267]: Use `.unique()`?

**Review by BugenZhao — APPROVED** [review]
LGTM! Surprised to find we didn't do this before.

**Pull request merged**